src/iOS.Autofill/CredentialProviderViewController.Passkeys.cs

```diff
@@ -25,52 +139,135 @@ private async Task ProvideCredentialWithoutUserInteractionAsync(ASPasskeyCredent
             }
 
-            // // TODO: Generate the credential Signature and Auth data accordingly
-            // var fido2AssertionResult = await _fido2AuthService.Value.GetAssertionAsync(new Bit.Core.Utilities.Fido2.Fido2AuthenticatorGetAssertionParams
-            // {
-            //     RpId = cipherView.Login.MainFido2Credential.RpId,
-            //     Counter = cipherView.Login.MainFido2Credential.Counter,
-            //     CredentialId = cipherView.Login.MainFido2Credential.CredentialId
-            // });
-
-            // CompleteAssertionRequest(new ASPasskeyAssertionCredential(
-            //     cipherView.Login.MainFido2Credential.UserHandle,
-            //     cipherView.Login.MainFido2Credential.RpId,
-            //     NSData.FromArray(fido2AssertionResult.Signature),
-            //     _context.PasskeyCredentialRequest?.ClientDataHash,
-            //     NSData.FromArray(fido2AssertionResult.AuthenticatorData),
-            //     cipherView.Login.MainFido2Credential.CredentialId
-            // ));
-        }
+            if (_context.PasskeyCredentialRequest is null)
+            {
+                OnProvidingCredentialException(new InvalidOperationException("Trying to complete assertion request without a PasskeyCredentialRequest"));
+                return;
+            }
 
-        public void CompleteAssertionRequest(ASPasskeyAssertionCredential assertionCredential)
-        {
-            if (_context == null)
+            try
+            {
+                // TODO: Add user verification and remove hardcoding on the user interface "userVerified"
+                var fido2AssertionResult = await _fido2AuthService.Value.GetAssertionAsync(new Bit.Core.Utilities.Fido2.Fido2AuthenticatorGetAssertionParams
+                {
+                    RpId = rpId,
+                    Hash = _context.PasskeyCredentialRequest.ClientDataHash.ToArray(),
+                    RequireUserVerification = _context.PasskeyCredentialRequest.UserVerificationPreference == "required",
+                    AllowCredentialDescriptorList = new Bit.Core.Utilities.Fido2.PublicKeyCredentialDescriptor[]
+                    {
+                        new Bit.Core.Utilities.Fido2.PublicKeyCredentialDescriptor
+                        {
+                            Id = credentialIdData.ToArray()
+                        }
+                    }
+                }, new Fido2GetAssertionUserInterface(cipherId, true, EnsureUnlockedVaultAsync, () => Task.FromResult(true)));
+
+                var selectedUserHandleData = fido2AssertionResult.SelectedCredential != null
+                    ? NSData.FromArray(fido2AssertionResult.SelectedCredential.UserHandle)
+                    : (NSData)userHandleData;
+
+                var selectedCredentialIdData = fido2AssertionResult.SelectedCredential != null
+                    ? NSData.FromArray(fido2AssertionResult.SelectedCredential.Id)
+                    : credentialIdData;
```

Comment:
question: why do we need to null-check these? Credential id will always be defined, and user handle will either be defined in both places or null in both places?

Reply:
@coroiu I did it at an early stage of this. If `SelectedCredential` must be non-null then should we just check if that's null we throw an exception?

Reply:
It is always defined so I think you can throw an exception (or just assume that it is defined)